### PR TITLE
feddyn implementation pytorch

### PR DESCRIPTION
**Pull request by GustavBaumgart**
Added FedDyn optimizer implementation (as described in the paper's pseudocode) compatible with pytorch only.

This involves an update function called by horizontal/trainer.py on the regularizers, and the trainer main.py files must send a deepcopy of the initial model (at each round) and the current model into the regularizer to get the algorithm's loss term.

### CODE REVIEW

**Review by jaemin-shin — CHANGES_REQUESTED** [review]
This is a neat and clean implementation of FedDyn, and I'm happy with this PR. One thing that I should ask is, do we have an example for FedDyn which we could run and test it, like medmnist_fedprox? It would be great if that is added and we could look at how it runs

File: lib/python/flame/optimizer/fedavg.py [hunk 1]
```diff
@@ -51,7 +51,8 @@ def do(self,
            cache: Cache,
            *,
            total: int = 0,
-           version: int = 0) -> ModelWeights:
+           version: int = 0,
+           **kwargs) -> ModelWeights:
```
[the remark this answers is not in the document]
Comment (GustavBaumgart) [reply lib/python/flame/optimizer/fedavg.py:55]: I believe the method "do" is not defined anywhere else (just inherited), so I basically changed the method signature in all instances (unless you find any that I left unchanged)

File: lib/python/flame/optimizer/feddyn.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        logger.debug("calling feddyn")
+        
+        num_trainers = kwargs['num_trainers']
+        num_selected = len(cache)
+        
+        # populate h_t
+        if self.h_t is None:
+            self.h_t = dict()
+            for k in base_weights:
+                self.h_t[k] = 0.0
```
[the remark this answers is not in the document]
Comment (GustavBaumgart) [reply lib/python/flame/optimizer/feddyn.py:81]: Actually, do you know what h_t is? I thought it meant something like "history," but the aggregation altogether is actually a bit strange.

File: lib/python/flame/optimizer/fedavg.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (GustavBaumgart) [reply lib/python/flame/optimizer/fedavg.py:55]: This will have to stay the same as before. Please let me know if you spot an error that could arise from this.

File: lib/python/flame/optimizer/regularizer/feddyn.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        
+        w_t_vector = None
+        for w_temp in w_t:
+            if not isinstance(w_t_vector, torch.Tensor):
+                w_t_vector = w_temp.reshape(-1)
+            else:
+                w_t_vector = torch.cat((w_t_vector, w_temp.reshape(-1)), 0)
+        
+        # grad term
+        inner_prod = torch.sum(self.prev_grad * w_vector)
```
[the remark this answers is not in the document]
Comment (GustavBaumgart) [reply lib/python/flame/optimizer/regularizer/feddyn.py:58]: no
